### ISSUE TITLE
Is there a way to know if the sqlite in use is threadsafe

### ISSUE THREAD

**Issue by teto** [ghaering/pysqlite#106]
The C api has `sqlite3_threadsafe()` to check for multithread support but I could not find an equivalent for the python API, is there an equivalent way to  retrieve the information ?
This would be for https://github.com/pimutils/khal/issues/507#issuecomment-248452490

**Comment by rogerbinns** [comment]
There are several issues being conflated here.  In the "olden" days SQLite 3 was not threadsafe.  It did not use any form of locking, so the caller had to do so.

Since then SQLite did add mutexes around every API call.  Those are present by default.  However people can choose to compile SQLite without them, and the sqlite3_threadsafe api is what tells you about that.  It is rare to compile SQLite without its default mutexes.

SQLite is still not entirely threadsafe though.  For example some APIs like the error ones return a pointer but any other thread could then cause a different error and that pointer is now invalid and de-referencing it could cause a crash, memory corruption or similar.  Consequently the error apis can only be accessed from multi-threaded code that is careful - eg https://github.com/rogerbinns/apsw/blob/master/src/util.c#L38 - which is something pysqlite doesn't do to my knowledge.  (I strongly campaigned to get the SQLite developers to do what every other platform does with errors, making them thread local which would address the issue.  They refused.)

Similarly if pysqlite did allow multi-threading, then it should protect against concurrent calls.  For example one thread could be processing SQLITE_ROW when a second thread calls close on the connection.  The first thread would then resume processing on invalid data structures.  Comments giving some idea of the work needed:  https://github.com/rogerbinns/apsw/blob/master/src/util.c#L7

In short, if you disabled pysqlite's single threaded mode, then doing multi-threading will lead to rare but horrifying crashes, corruption, wrong data or messages, and other unexpected behaviour.

A recommended pattern is to open separate connections in each thread.  That will work fine with all versions of pysqlite and SQLite no matter how they were compiled.  You will need to ensure that transactions are bounded correctly since they will only be visible to other connections once committed.

**Comment by domenkozar** [comment]
@rogerbinns what about using `RLock` for each thread?

**Comment by rogerbinns** [comment]
@domenkozar if you want to be safe, then do as above - use a connection per thread.  If you want to use the same connection across threads then pysqlite needs to be updated.  In any event disabling the default builtin locking of SQLite in favour of a home grown solution is not worth it (you only need to one mistake to have a very hard to debug problem).
